k8s/watchers: fix data race in (*K8sWatcher).addK8sServiceV1

tklauser opened the pull request:
Access to k.podStore must be protected by k.podStoreMU.

Closes #13603

Fixes: e7bb8a7eadb5 ("k8s/cilium Event handlers and processing logic for LRPs")

tklauser committed:
k8s/watchers: fix data race in (*K8sWatcher).addK8sServiceV1

Access to k.podStore must be protected by k.podStoreMU.

Closes #13603

Fixes: e7bb8a7eadb5 ("k8s/cilium Event handlers and processing logic for LRPs")
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

aanm reviewed (changes requested):
@tklauser I'm fine with the changes but this should not be the real fix. This code architecture should be done differently as well.

@aditighag `k.podStore` should not be passed for every k8s service add, we should register a store in https://github.com/cilium/cilium/blob/e91f803c360484dc109eb0d18badc1afa92d457f/daemon/cmd/daemon.go#L367 and use it as follows:

```patch
diff --git a/daemon/cmd/daemon.go b/daemon/cmd/daemon.go
index 8c334acaa..faae4336b 100644
--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -365,6 +365,7 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
        )
 
        d.redirectPolicyManager.RegisterSvcCache(&d.k8sWatcher.K8sSvcCache)
+       d.redirectPolicyManager.RegisterGetStores(&d.k8sWatcher)
 
        bootstrapStats.daemonInit.End(true)
 
diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index 5213fcb30..bfce4b013 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -65,6 +65,8 @@ type Manager struct {
 
        svcCache svcCache
 
+       storeGetter StoreGetter
+
        // Mutex to protect against concurrent access to the maps
        mutex lock.Mutex
 
@@ -95,6 +97,14 @@ func (rpm *Manager) RegisterSvcCache(cache svcCache) {
        rpm.svcCache = cache
 }
 
+type StoreGetter interface {
+       GetStore(name string) cache.Store
+}
+
+func (rpm *Manager) RegisterGetStores(sg StoreGetter) {
+       rpm.storeGetter = sg
+}
+
 // Event handlers
 
 // AddRedirectPolicy parses the given local redirect policy config, and updates
@@ -503,12 +513,13 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 }
 
 // Returns a slice of endpoint pods metadata that are selected by the given policy config.
-func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Store) []*podMetadata {
+func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig) []*podMetadata {
        var (
                retPods []*podMetadata
                podData *podMetadata
        )
 
+       podStore := rpm.storeGetter.GetStore("pod")
        for _, podItem := range podStore.List() {
                pod, ok := podItem.(*slimcorev1.Pod)
                if !ok || !config.checkNamespace(pod.GetNamespace()) {

```

tklauser committed:
redirectpolicy: use StoreGetter to get pod store

Dont' pass k.podStore for every k8s service add. Instead, use a
StoreGetter to get the pod store.

Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

christarazi commented on pkg/redirectpolicy/manager_test.go:
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/cilium/cilium/pkg/k8s/utils"
 	lb "github.com/cilium/cilium/pkg/loadbalancer"
 	"github.com/cilium/cilium/pkg/policy/api"
+	"k8s.io/client-go/tools/cache"
```
Nit: import ordering

christarazi reviewed (approved):
Minor nit
